### PR TITLE
fix: resolve 6 high-impact agentic workflow issues (body limit, timeouts, tool_choice, system role, schema)

### LINKED ISSUE #1529: 【Request Body Limit】Large /v1/responses requests are truncated by Next proxy 10MB body limit and reported as "Invalid JSON body"
## 问题描述

当通过 9Router 的 `/v1/responses` 发送较大的 OpenAI Responses API 请求时，请求体超过约 10MB 后会失败，并返回：

```json
{"error":{"message":"Invalid JSON body","type":"invalid_request_error","code":"bad_request"}}
```

控制台同时出现 Next 的提示：

```text
Request body exceeded 10MB for /v1/responses. Only the first 10MB will be available unless configured.
```

这说明请求并不是原始 JSON 语法错误，而是 Next proxy 默认 `proxyClientMaxBodySize` 为 10MB，超过后只把前 10MB body 交给后续 route handler。之后 `request.json()` 读到的是被截断的 JSON，所以被 9Router 包装成 `Invalid JSON body`。

## 影响

这会影响长上下文、多图片、base64 image input、以及本地 agent 客户端通过 `/v1/responses` 发送的大请求。

OpenAI 的图片输入公开文档中，每次请求 total payload 上限可到 512MB。因此 9Router 当前 10MB 的入口限制明显偏低，容易在模型上下文还没耗尽前先被本地路由层截断。

相关文档：

- Next.js `proxyClientMaxBodySize`: https://nextjs.org/docs/app/api-reference/config/next-config-js/proxyClientMaxBodySize
- OpenAI image input requirements: https://developers.openai.com/api/docs/guides/images-vision

## 当前行为

- 小于 10MB 的合法 JSON 可以继续进入后续处理，例如鉴权。
- 约 10MB 或更大的合法 JSON 会返回 `Invalid JSON body`。
- 错误信息会误导用户以为是客户端发了坏 JSON，而不是 body 被 proxy 截断。

## 期望行为

建议修复方向：

1. 为 Next 配置 `experimental.proxyClientMaxBodySize`，默认值建议提高到 `128mb`。
2. 提供环境变量或 CLI 配置，例如 `NINEROUTER_PROXY_CLIENT_MAX_BODY_SIZE=256mb` 或 `512mb`，方便用户按机器内存和使用场景调整。
3. 超限时返回更明确的错误，例如 `413 Payload Too Large` 或提示 `Request body exceeded proxyClientMaxBodySize`，不要继续泛化成 `Invalid JSON body`。
4. 长期更优方案：让 `/v1` / `/api/v1` 这类 LLM API 请求不要经过会 clone request body 的 Next proxy，或者把认证逻辑下沉到 route handler，避免中间层复制大请求体。

## 推荐默认值

建议默认值为 `128mb`，不是直接改到 `512mb`。

原因：

- `64mb` 对长文本和少量截图通常够用，但单张 50MB 图片经过 base64 后会膨胀到约 67MB，再加 JSON 包装可能超过。
- `128mb` 能覆盖常见长上下文、多截图、单张较大图片场景。
- `512mb` 虽然贴近 OpenAI 图片输入上限，但 Next proxy clone body + `request.json()` 解析会带来明显内存占用，作为默认值对本地路由器偏激进。
- 高阶用户仍可通过环境变量手动调到 `256mb` 或 `512mb`。

## 复现方式

以下命令只发送 dummy text，不包含真实会话或密钥。

小于 10MB 时，请求体可以被解析，之后进入鉴权逻辑：

```bash
node -e 'process.stdout.write(JSON.stringify({model:"gpt-5.5",input:[{type:"message",role:"user",content:[{type:"input_text",text:"x".repeat(9*1024*1024)}]}],stream:true}))' \
  | curl -sS -i http://127.0.0.1:20128/v1/responses \
      -H 'Content-Type: application/json' \
      --data-binary @-
```

约 10MB 或更大时，合法 JSON 被截断，返回 `Invalid JSON body`：

```bash
node -e 'process.stdout.write(JSON.stringify({model:"gpt-5.5",input:[{type:"message",role:"user",content:[{type:"input_text",text:"x".repeat(10*1024*1024)}]}],stream:true}))' \
  | curl -sS -i http://127.0.0.1:20128/v1/responses \
      -H 'Content-Type: application/json' \
      --data-binary @-
```

## 环境

- 9Router version: `0.4.62`
- Endpoint: `/v1/responses`
- Runtime: npm global package / Next standalone
- Observed packed config: `experimental.proxyClientMaxBodySize: 10485760`

### LINKED ISSUE #1557: Make STREAM_STALL_TIMEOUT_MS and FETCH_CONNECT_TIMEOUT_MS configurable via Env Vars or UI
Currently, `STREAM_STALL_TIMEOUT_MS` and `FETCH_CONNECT_TIMEOUT_MS` appear to be hardcoded values. When routing to slow-streaming endpoints (such as performing heavy tasks), these default limits can lead to premature timeout errors and dropped connections.  Making these variables configurable would provide much needed flexibility for slower endpoints.

### LINKED ISSUE #1564: Claude Code fails with Antigravity provider on all models due to invalid tool schema
Hello, I want to report an issue with Claude Code when using the Antigravity provider.

Claude Code cannot be used with Antigravity on any model. Every request fails with the same API error, for example:

API Error: 400 [antigravity/gemini-3.5-flash-extra-low] [400]: {
  "error": {
    "code": 400,
    "message": "Invalid value at 'request.tools[0].function_declarations[28].parameters.properties[6].value' (type.googleapis.com/google.cloud.aiplatform.master.Schema), \"object\"",
    "status": "INVALID_ARGUMENT",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.BadRequest",
        "fieldViolations": [
          {
            "field": "request.tools[0].function_declarations[28].parameters.properties[6].value",
            "description": "Invalid value at 'request.tools[0].function_declarations[28].parameters.properties[6].value' (type.googleapis.com/google.cloud.aiplatform.master.Schema), \"object\""
          }
        ]
      }
    ]
  }
}

This happens consistently across all Antigravity models I tried, so it looks like Claude Code is sending a tool/function schema that is not accepted by Google Vertex AI / Gemini-compatible schema validation.

It seems the problematic field is:

request.tools[0].function_declarations[28].parameters.properties[6].value

The value appears to be "object", but the provider rejects it as an invalid schema value.

Expected behavior:

Claude Code should work normally with the Antigravity provider, or the tool/function schema should be converted into a format compatible with the provider.

Actual behavior:

Claude Code immediately fails with API Error 400 before the model can respond.

Notes:

This does not seem to be model-specific, because the same error appears on multiple Antigravity models.

Please check whether Claude Code is generating an invalid function/tool schema for Gemini/Vertex-compatible providers.

Thank you.

### LINKED ISSUE #1580: [Bug] role 'system' is not supported on this model — cc/claude-sonnet-4-6 always fails with 400
### Version
`v0.4.66`

### Description
When using a Combo with `cc/claude-sonnet-4-6` as the primary model, every request fails with HTTP 400 and the error:

```
{"type":"error","error":{"type":"invalid_request_error","message":"role 'system' is not supported on this model"}}
```

The request eventually succeeds via fallback to `kr/claude-sonnet-4.6` (Kiro), so functionality is not completely broken — but the primary provider (`cc`) always fails, causing unnecessary latency and exhausting fallback slots.

### Steps to Reproduce
1. Create a Combo (e.g. `MyComboSonnet`) with `cc/claude-sonnet-4-6` as the first model
2. Send any request through the combo (e.g. from Claude Code)
3. Observe the 400 error in the logs

### Logs
```
[ERROR] [400]: {"type":"error","error":{"type":"invalid_request_error","message":"role 'system' is not supported on this model"},"request_id":"req_011CbaHP3ga95YE6zRSHWKwm"}
[ERROR] [400]: {"type":"error","error":{"type":"invalid_request_error","message":"role 'system' is not supported on this model"},"request_id":"req_011CbaHPCg4aY1yCtywSMBf9"}
```
After 2 failures, 9router falls back to `kr/claude-sonnet-4.6` (Kiro) which succeeds.

### Root Cause (suspected)
The Anthropic API does **not** accept `system` as a role inside the `messages` array. It must be sent as a **top-level `system` field** in the request body. It appears that 9router's format translation for the `cc` (Claude Code) provider may be passing `system` as a message role instead of a top-level field, causing the 400 error.

See Anthropic API docs: https://docs.anthropic.com/en/api/messages

### Expected Behavior
Requests to `cc/claude-sonnet-4-6` should succeed without error. The `system` prompt should be properly translated to a top-level field before being forwarded to the Anthropic API.

### Actual Behavior
All requests to `cc/claude-sonnet-4-6` fail with 400. The combo falls back to the next provider every time.

### Environment
- 9router version: `v0.4.66`
- Client tool: Claude Code
- Combo: `MyComboSonnet` (strategy: fallback, sticky: 1)
- Models in combo: `cc/claude-sonnet-4-6` → `kr/claude-sonnet-4.6`

### Workaround
Using `kr/claude-sonnet-4.6` (Kiro) directly as the primary model works fine.

### LINKED ISSUE #1592: [Bug] Forced tool_choice fails (400) on cc/ Claude route — type not translated + tool_choice.name not suffixed
## Summary

When a Claude account is connected to 9router via OAuth (provider code `cc`), any OpenAI-compatible client that **forces a specific tool** via `tool_choice` gets a 400. Two separate bugs in the OpenAI→Claude request translation, surfacing one at a time.

## Setup to reproduce

- 9router `0.4.66` running locally (proxy on `http://localhost:20128`).
- A **Claude account added via OAuth** — *Add provider → Claude → sign in* (a Claude Pro/Max login; stores an `sk-ant-oat…` token). 9router addresses it by the provider code **`cc`** ("Claude Code"), so the model id is **`cc/claude-sonnet-4-6`**. This OAuth path is what gates 9router's Claude tool "cloaking" (`applyCloaking` / `cloakClaudeTools`), which is why bug 2 only appears here.
- Client: any OpenAI-compatible client pointed at `…/v1/chat/completions` that forces a tool. Reproduced with `pi` (`@earendil-works/pi-coding-agent`, `api=openai-completions`), which pins the `todo_write` tool via `tool_choice`.

## Steps

`POST http://localhost:20128/v1/chat/completions`

```json
{
  "model": "cc/claude-sonnet-4-6",
  "messages": [{ "role": "user", "content": "add todo buy milk" }],
  "tools": [{ "type": "function", "function": {
    "name": "todo_write",
    "parameters": { "type": "object", "properties": { "items": { "type": "array", "items": { "type": "string" } } } }
  }}],
  "tool_choice": { "type": "function", "function": { "name": "todo_write" } }
}
```

A credential-free, deterministic repro (no token, no network — runs the translator functions directly) is in the comment below.

## Expected

The forced tool is honored and the request returns 200 with a `todo_write` tool call.

## Actual

400 from Claude, one error at a time:

```
tool_choice: Input tag 'function' does not match expected tags: 'auto','any','tool','none'   ← bug 1
Tool 'todo_write' not found in provided tools                                                ← bug 2 (after bug 1 is fixed)
```

## Root cause

1. **`tool_choice.type` not translated** — `convertOpenAIToolChoice()` (`open-sse/translator/request/openai-to-claude.js`) returns any object with a `.type` verbatim before the `.function` branch. OpenAI's `{type:"function",function:{name}}` carries `.type === "function"`, which Claude rejects.
2. **`tool_choice.name` not cloaked** — `cloakClaudeTools()` (`open-sse/utils/claudeCloaking.js`) renames client tools with `CLAUDE_TOOL_SUFFIX` (`_ide`) but leaves `tool_choice.name` unchanged, so a forced `{type:"tool",name:"todo_write"}` no longer matches the renamed `todo_write_ide`.

## Impact

Any client that pins a tool (`todo_write`, structured-output forcing, agent frameworks) is broken on the OAuth Claude (`cc`) route. `tool_choice: "auto"`/`"none"`/`"required"` and no-`tool_choice` requests are unaffected.

## Fix

PR #1593 — translates the forced-tool shape, suffixes `tool_choice.name` to match the cloaked tools, and hardens both paths against malformed/unknown input. Includes unit coverage.

### PR DESCRIPTION
**Pull request by GodrezJr2**
## Summary

Six bugs that break or degrade daily agentic workflows (Claude Code + Codex + other models via 9router). Each fix is isolated, tested before/after, and references the relevant issue.

---

### Fix 1 — 10MB body limit silently truncates long context sessions (closes #1529 #1572)

**File:** `next.config.mjs`

Next.js's default 10MB `proxyClientMaxBodySize` truncates requests silently. Long agentic sessions (many tool calls + conversation history) easily exceed this — the proxy logs success but the downstream provider receives malformed JSON.

```js
experimental: {
  proxyClientMaxBodySize: process.env.NEXT_MAX_BODY_SIZE || "128mb"
}
```

**Test:** `config.experimental.proxyClientMaxBodySize === "128mb"` ✓ env-override ✓

---

### Fix 2 — Stream stall / connect timeouts hardcoded (closes #1557)

**File:** `open-sse/config/runtimeConfig.js`

`STREAM_STALL_TIMEOUT_MS` (30s) and `FETCH_CONNECT_TIMEOUT_MS` (20s) are hardcoded. Slow-streaming models (Kiro, large reasoning models) hit the stall timeout mid-task. Now configurable via env vars with a minimum floor guard.

**Test:** env `STREAM_STALL_TIMEOUT_MS=120000` → constant = 120000 ✓

---

### Fix 3 — Forced `tool_choice` fails 400 on `cc/` Claude OAuth route (closes #1592)

**Files:** `open-sse/translator/request/openai-to-claude.js`, `open-sse/utils/claudeCloaking.js`

Two bugs:
1. `convertOpenAIToolChoice` returned `{type:"function", ...}` verbatim before checking the `.function` branch → Claude rejected `type:"function"` 
2. `cloakClaudeTools` renamed tools with `_ide` suffix but left `tool_choice.name` unchanged → forced tool name no longer matched any declaration

```
Before: tool_choice = {"type":"function","function":{"name":"todo_write"}}  → 400
After:  tool_choice = {"type":"tool","name":"todo_write_ide"}               → 200
```

**Test:** exact repro from issue comment — both bug assertions = false ✓

---

### Fix 4 — `cc/claude-sonnet-4-6` always 400: `role 'system' not supported` (closes #1580)

**File:** `open-sse/translator/helpers/claudeHelper.js`

Anthropic's API rejects `role:"system"` in `messages[]` — system content must be the top-level `system` field. `prepareClaudeRequest` now filters any system/developer-role messages from `messages[]` and promotes their text to the top-level `system` field. Guards the passthrough path (Claude Code → cc/) and any translation edge cases.

**Test:** body with `role:"system"` in messages → extracted to `system` field, messages clean ✓

---

### Ref #1382 — DeepSeek 0-output silent fail with tool-heavy requests (partial)

**File:** `open-sse/handlers/chatCore/streamingHandler.js`

Full combo-fallback for streaming requires buffering the response before forwarding — a larger architectural change. As a first step, `onStreamComplete` now emits a clear `[WARN][EMPTY-RESPONSE]` log when `out_tokens < 50` AND `tools.length >= 5`. This lets operators identify which provider to remove from tool-heavy combos.

---

### Fix 5 — Claude Code fails with Antigravity/Gemini: invalid tool schema (closes #1564)

**File:** `open-sse/translator/helpers/geminiHelper.js`

Claude Code tool schemas sometimes define properties as plain type strings (e.g. `"field": "object"` instead of `"field": {"type":"object","properties":{}}`). Vertex AI rejects these with a 400. New `fixInvalidPropertyValues` phase in `cleanJSONSchemaForAntigravity` converts string-shorthand property values to proper schema objects.

```
Before: properties.options = "object"         → Vertex 400
After:  properties.options = {type:"object"}  → 200
```

**Test:** schema with `"command":"string"`, `"options":"object"`, `"count":"integer"` — all converted correctly ✓

---

## Files changed

| File | Issue |
|------|-------|
| `next.config.mjs` | #1529 #1572 |
| `open-sse/config/runtimeConfig.js` | #1557 |
| `open-sse/translator/request/openai-to-claude.js` | #1592 bug 1 |
| `open-sse/utils/claudeCloaking.js` | #1592 bug 2 |
| `open-sse/translator/helpers/claudeHelper.js` | #1580 |
| `open-sse/handlers/chatCore/streamingHandler.js` | #1382 (partial) |
| `open-sse/translator/helpers/geminiHelper.js` | #1564 |

All fixes verified with unit-level node tests before/after. No new vitest failures introduced.

### CODE REVIEW

File: open-sse/handlers/chatCore/streamingHandler.js [hunk 1]
```diff
@@ -93,6 +93,17 @@ export function buildOnStreamComplete({ provider, model, connectionId, apiKey, r
     });
 
     saveUsageStats({ provider, model, tokens: usage, connectionId, apiKey, endpoint: clientRawRequest?.endpoint, label: "STREAM USAGE" });
+
```
Comment (tonydwb) [inline open-sse/handlers/chatCore/streamingHandler.js:101]: 🔴 **Stale debug artifact**: Remove or gate this `console.warn` with ANSI color escape codes. It will emit on every near-empty tool-heavy provider response in production. Use the structured logger (`dbg`/log) or remove it entirely.

**Review by tonydwb — CHANGES_REQUESTED** [review]
## Code Review Summary

**Verdict: Changes Requested**

Good scope overall, but there is a debug artifact that should be removed before merge.

### 🔴 Critical
- `open-sse/handlers/chatCore/streamingHandler.js:96` — **Stale debug `console.warn` left in production code**. The `[WARN] [EMPTY-RESPONSE]` warning with ANSI color codes was added for diagnostics and should be either (a) gated behind a debug flag, (b) removed, or (c) routed through the existing logging system. Shipping raw colored console output from a streaming handler is a debug artifact and will emit in server logs on every near-empty tool-heavy response.

### ✅ Looks Good
- Configurable body limit, stream stall timeout, and fetch connect timeout are clean.
- System/developer role promotion in claudeHelper is a solid safety net for passthrough.
- `fixInvalidPropertyValues` in geminiHelper handles a real Gemini/Vertex edge case cleanly.

---
*Reviewed by Hermes Agent (cron)*

**Comment by GodrezJr2** [comment]
Conflict resolved and branch updated onto current upstream master. The prior ANSI log review item was also addressed in the follow-up commit. Ready for re-review when you have time.

**Comment by GodrezJr2** [comment]
@decolua friendly bump — conflict resolved, rebased on current master, CLEAN and mergeable. Earlier ANSI-log review item was addressed. Bundles 6 high-impact agentic fixes (body limit, stream/connect timeouts, Claude tool_choice, system-role, Gemini schema). Could you review/merge when you have a moment? Thanks.
